### ISSUE TITLE
Issue with TFLite Inference for Custom Dataset Detection

### ISSUE THREAD

**Issue by wamiqraza** [ultralytics/ultralytics#5119]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict, Export

### Bug

I have few few issue with tflite infrence:

I am writing to seek assistance with a specific issue I am encountering while performing object detection using a custom dataset and TFLite with float32.

The problem I am facing is that, despite using TFLite for inference, the results consistently yield predictions of the same class for all detected objects, regardless of their actual class. I have attached a screenshot for your reference. Also the bounding boxes are not around the actual object

To provide context, I am utilizing the code provided in the linked resource [Here](https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-ONNXRuntime/main.py). I have followed the steps outlined, but unfortunately, I'm still encountering this issue.

I would greatly appreciate any guidance or insights you can offer in resolving this matter. 

Moreover, the other issue with pretrained model facing this issue: [Here](https://github.com/ultralytics/ultralytics/issues/992#issuecomment-1737588934) this for was debagging for above task and face this error.

![output](https://github.com/ultralytics/ultralytics/assets/22597483/d2fe805a-db6a-4c97-86cd-9669bacb049c)
![output1](https://github.com/ultralytics/ultralytics/assets/22597483/d0733adb-4cb3-4259-9efb-b743660dc7c4)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@razacode thank you for the detailed articulation of your issue. There are a few aspects that we can investigate to resolve this.

For the problem where all detections are being classified as the same class, it often indicates an issue with the labeling during training or the interpretation of the classes during prediction. Can you please confirm that the correct class labels are being used during both training and inference stages?

Regarding the issue of bounding boxes not being correctly positioned around objects, this typically suggests that the model has either not been adequately trained or the anchor box dimensions are not optimal for the object sizes in your custom dataset. I would advise observing the training progress and the associated measures like loss and mean average precision during your model training phase. 

Finally, the error with the pre-trained model that you mentioned appears not to be related directly to the problem in discussion, as it looks to be an issue with your environment setup. Our team would need further details on that, but I recommend isolating and troubleshooting that error separately.

Remember that effective training of deeper networks like YOLOv8 requires ample labeled data, thorough hyperparameter tuning, and sufficient compute resources.

As you've indicated that you are willing to contribute a PR, it would be helpful if you could investigate this behavior further, possibly trying to reproduce it using different environments or configurations, and sharing your results in the PR.

Your collaboration working on this issue is greatly appreciated and vital for the improvements within the YOLOv8 and the wider community.

**Comment by razacode** [comment]
> @razacode thank you for the detailed articulation of your issue. There are a few aspects that we can investigate to resolve this.
> 
> For the problem where all detections are being classified as the same class, it often indicates an issue with the labeling during training or the interpretation of the classes during prediction. Can you please confirm that the correct class labels are being used during both training and inference stages?
> 
> Regarding the issue of bounding boxes not being correctly positioned around objects, this typically suggests that the model has either not been adequately trained or the anchor box dimensions are not optimal for the object sizes in your custom dataset. I would advise observing the training progress and the associated measures like loss and mean average precision during your model training phase.
> 
> Finally, the error with the pre-trained model that you mentioned appears not to be related directly to the problem in discussion, as it looks to be an issue with your environment setup. Our team would need further details on that, but I recommend isolating and troubleshooting that error separately.
> 
> Remember that effective training of deeper networks like YOLOv8 requires ample labeled data, thorough hyperparameter tuning, and sufficient compute resources.
> 
> As you've indicated that you are willing to contribute a PR, it would be helpful if you could investigate this behavior further, possibly trying to reproduce it using different environments or configurations, and sharing your results in the PR.
> 
> Your collaboration working on this issue is greatly appreciated and vital for the improvements within the YOLOv8 and the wider community.

**Comment by razacode** [comment]
@glenn-jocher thank you very much for your timely reply.

With regards to labeling I have label every image correctly and to make sure this I have run below command to do prediction on .pt model:
`yolo detect mode=predict model='./runs/detect/train/weights/best.pt' conf=0.25 source='test'`

And it gives the correct bounding box against a particular class **image** is attached for your reference.

![boat-summer-sea-ship-sail-travel-1634811](https://github.com/ultralytics/ultralytics/assets/22597483/b3d568f6-0401-49d6-a2f5-dbd2abbf8bfe)

With .onnx model also it draw a bounding box with `yolo detect mode=predict model='./runs/detect/train/weights/best.onnx' conf=0.25 source='test'`

I have figure it out and and to contribute a PR I can add the code that does the FLOAT32 tflite prediction using `from tflite_runtime import interpreter as tflite`. Now I have tested my code on both pretrained and custom trained model and it work.

**Comment by glenn-jocher** [comment]
Hi @razacode, 

It's great to hear that you have meticulously double-checked your labeling and run prediction using both the .pt and .onnx models, also that you have found a way to make the FLOAT32 tflite prediction work with the tflite_runtime as you shared.

Your experience and progress is valuable to the community. If you could contribute to the repository with the PR illustrating what you discovered and how you solved your problem, it will undoubtedly be beneficial for others who may encounter a similar issue in the future.

Thank you for your commitment to improving and enriching the repository. We eagerly look forward to your PR!

**Comment by razacode** [comment]
Hi @glenn-jocher,

Sure, really appreatiate your timely responce and I will work on it these days then will do a PR. With regard to closing this issue will close after PR as we can refer user with such issue to the correspondence section of code.

Thank you

**Comment by glenn-jocher** [comment]
Hi @razacode,

That sounds like a great plan! We definitely appreciate you working on this issue and looking forward to your PR. Your efforts towards improving the efficiency and accuracy of the model are truly valuable. Feel free to reach out if you have any queries in the process. We are here to support you.

As for closing this issue, your suggestion sounds perfect. Keeping this issue open until the PR is submitted will indeed be helpful for other users facing a similar problem.

Thank you for your proactive contribution and let us know how we can further assist you.

**Comment by wamiqraza** [comment]
Hi @glenn-jocher 

As I have mentioned in my early conversation I have worked on it this weekend. Now I am ready to push the tflite implementation to the repo what will be the procedure to contribute as I have clone and pull the repository and added file into my local repository. Upon git push origin main it show me this message remote: **Permission to ultralytics/ultralytics.git denied to wamiqraza** .
Moreover, I would like to connect via LinkedIn https://it.linkedin.com/in/wamiq-raza to chat there in personal about the contribution.

Thank you very much looking forward

**Comment by glenn-jocher** [comment]
That's great news that you have an update! Please submit a Pull Request to the repo. You can do this from your fork on GitHub. 

For a guide on submitting PRs please see https://docs.ultralytics.com/help/contributing

**Comment by glenn-jocher** [comment]
Hi @wamiqraza,

Thanks for submitting your PR! Our team will review the changes and provide feedback. Good work on adding the float32 tflite file. We might need more time to determine if it's beneficial to all users. Your work is very valuable and we appreciate your understanding.

Your suggested edits in the readme, script, and output image will be reviewed in depth. Thanks again for your hard work and contribution to the YOLOv8 project!

**Comment by wamiqraza** [comment]
Hi @glenn-jocher,

You are welcome and looking forward for the review team comments. One thing is for sure with this contribution the issue with respect to tflite will get reduce as after thoroughly reading most the issue with tflite I come up with this to solve it for the user. 

Thanks again

**Comment by glenn-jocher** [comment]
Hello @wamiqraza,

We're grateful for your initiative towards resolving the issues with tflite. There's no doubt that your hard work and diligent efforts will contribute fantastically to the YOLOv8 community.

The review team will follow up on your PR soon. We are all looking forward to implementing your work and further advancing the functionality of YOLOv8.

Once again, thank you for your significant contribution. It is clear that your updates will have a meaningful impact for our users.

**Comment by wamiqraza** [comment]
Hi @glenn-jocher,

Happy new year!

I have finalise the int8 tflite implementation and submitted for PR. Moreover, script can be found here for the reader: https://github.com/wamiqraza/ultralytics/tree/yolov8-int8-tflite

Thanks looking forwards for your comments

**Comment by glenn-jocher** [comment]
Happy New Year @wamiqraza!

Thank you for submitting the PR for the int8 tflite implementation. We'll review it and get back to you with comments. Your contributions are greatly appreciated!

**Comment by wamiqraza** [comment]
Issues are rectified, refer to: [PR](https://github.com/ultralytics/ultralytics/pull/7317)
or implementation: [script ](https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-OpenCV-int8-tflite-Python)

**Comment by glenn-jocher** [comment]
@wamiqraza thank you for the update. We'll take a look at the PR and the implementation script. Your efforts in rectifying the issues are much appreciated! 🚀
